### PR TITLE
Use NSDiffableDataSource for iOS 13 or greater

### PR DESCRIPTION
**Pull request by jasonkemp9099**
To workaround crashes in UITableView with large change sets, use UITableViewDiffableDataSource when available (iOS 13 or greater)

Created an Impl protocol, with two implementations:

1. Original Functional Table Data algorithm
2. UITableViewDiffableDataSource-backed implementation for iOS 13 and up.

Then just followed the consequences of that. Any other changes are due to getting it working on IOS 13 at runtime. Tests with data exhibiting the crash with the original table data algorithm pass with diffable data sources.

### CODE REVIEW

File: Sources/FunctionalTableData/TableSectionHeaderFooter.swift [hunk 1]
```diff
@@ -27,6 +27,10 @@ public protocol TableHeaderFooterStateType: Equatable {
 }
 
 public struct TableSectionHeaderFooter<ViewType: UIView, Layout: TableItemLayout, S: TableHeaderFooterStateType>: TableHeaderFooterConfigType {
+	public func dequeueCell(from tableView: UITableView, at indexPath: IndexPath) -> UITableViewCell {
+		return UITableViewCell()
+	}
```
Comment (Campbell-Scott) [inline Sources/FunctionalTableData/TableSectionHeaderFooter.swift:32]: Why is this needed for `TableHeaderFooterConfigType`? `AnyCellConfigType` conformance?
Is there any consequence to returning an empty cell here - are we ever expecting this to be called on this type?

Would it make sense to have an `AnyTableHeaderConfigType` that leans on `func dequeueHeaderFooter(from tableView: UITableView) -> UITableViewHeaderFooterView?` instead?

Comment (jasonkemp9099) [reply Sources/FunctionalTableData/TableSectionHeaderFooter.swift:32]: DIffable data sources fails with a crash because of a default implementation of `dequeueCell()` that returned `UITableViewCell.cellIdentifier`. Deleting that is in the diff here as well. I suspect it has to do with default implementations in protocol extensions and inheritance chains, sometimes those aren't predictable.

From what I can tell, `TableHeaderFooterConfigType` is also a `CellConfigType`, so that dequeueCell is necessary. I can't think of a scenario where we call that when using the TableHeaderFooterConfigType as a TableHeaderFooterConfigType, but if someone chooses to also use it as a CellConfigType and doesn't do the necessary work in their implementation, that could spell trouble. That seems remote enough as a possibility to do this.

**Comment by josephroquedev** [comment]
NSDiffableDataSource is dropping cells if they have the same ID/content as a cell in a different section.

| With same ID / different content | With same ID / same content |
|---|---|
| Simulator Screen Shot - iPhone 12 Pro - 2021-03-17 at 10 52 59 | Simulator Screen Shot - iPhone 12 Pro - 2021-03-17 at 10 53 13 |

Only the latter cell with identical content is being displayed. Also getting the following message in the console: `[UIDiffableDataSource] Warning: 2 inserted identifier(s) already present; existing items will be moved into place for this current insertion. Please note this will impact performance if items are not unique when inserted.`

If I change the content inside the cell or make the cell IDs globally unique, then they both render. Not an issue with classic diffing strategy

**Comment by jasonkemp9099** [comment]
> NSDiffableDataSource is dropping cells if they have the same ID/content as a cell in a different section.

Nice find. Has to do with my Hashable implementations, I bet. I should make a snapshot test for this too.

File: Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift [hunk 2]
```diff
@@ -57,4 +57,248 @@ extension FunctionalTableData {
...
+		
+		private func dumpDebugInfoForChanges(_ changes: TableSectionChangeSet, previousSections: [TableSection], visibleIndexPaths: [IndexPath], exceptionReason: String?, exceptionUserInfo: [AnyHashable: Any]?) {
+			guard let exceptionHandler = FunctionalTableData.exceptionHandler else { return }
+			let exception = Exception(name: name, newSections: dataSource.data.sections, oldSections: previousSections, changes: changes, visible: visibleIndexPaths, viewFrame: tableView?.frame ?? .zero, reason: exceptionReason, userInfo: exceptionUserInfo)
+			exceptionHandler.handle(exception: exception)
+		}
+	}
+	
+	@available(iOS 13.0, *)
+	class DiffableDataSourceFunctionalTableDataImpl: FunctionalTableDataImpl {
```
[the remark this answers is not in the document]
Comment (jasonkemp9099) [reply Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift:260]: Yes, duplicated the test specifically for perf tests and any others I could find that used a `FunctionalTableData` instance, now that we make it opt-in to diffable data sources.

File: Sources/FunctionalTableData/CellConfigType.swift [hunk 3]
```diff
@@ -63,3 +63,73 @@ public extension CellConfigType {
...
+	private let sectionKey: String
+	
+	init(_ base: CellConfigType, sectionKey: String) {
+		self.base = base
+		self.sectionKey = sectionKey
+	}
+	
+	func hash(into hasher: inout Hasher) {
+		hasher.combine(key)
+		hasher.combine(accessibility)
```
[the remark this answers is not in the document]
Comment (jasonkemp9099) [reply Sources/FunctionalTableData/CellConfigType.swift:105]: I'll pull that, I think. It came before adding `sectionKey`, the ol' kitchen sink approach. :)
I think using AnyCellConfigType to tie section and configType together for both hashable and equality testing is the trick to using diffable data sources here, after the first round of testing and doing some digging online.

File: Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift [hunk 4]
```diff
@@ -57,4 +57,248 @@ extension FunctionalTableData {
...
+			tableView?.registerCellsForSections(localSections)
+			cellStyler.data.sections = localSections
+			
+			var snapshot = NSDiffableDataSourceSnapshot<TableSection, AnyCellConfigType>()
+			for newSection in localSections {
+				snapshot.appendSections([newSection])
+				snapshot.appendItems(newSection.anyRows)
+			}
+			
+			datasource?.apply(snapshot, animatingDifferences: false, completion: completion)
```
[the remark this answers is not in the document]
Comment (jasonkemp9099) [reply Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift:295]: OK, figured this out. 

UITableViewDiffableDataSource offers `defaultRowAnimation` to customize the animation when a snapshot is applied. We can't customize down to each operation like previously. We pick `.fade` for insert and delete, and `.automatic` for reload. Tried both. Automatic is the default for any UITableView. It's much more prevalent using diffable data sources to my eyes. `.fade` is a bit nicer.

I'll set `defaultRowAnimation` to `animations.rows.insert` here for now.

[the remark this answers is not in the document]
Comment (jasonkemp9099) [reply Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift]: Another follow up! 😄

Scott posted some pretty compelling evidence in the issue channel in the form of videos to preferring `.none` as the default row animation

File: Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift [hunk 5]
```diff
@@ -57,4 +57,248 @@ extension FunctionalTableData {
...
+			guard let exceptionHandler = FunctionalTableData.exceptionHandler else { return }
+			let exception = Exception(name: name, newSections: dataSource.data.sections, oldSections: previousSections, changes: changes, visible: visibleIndexPaths, viewFrame: tableView?.frame ?? .zero, reason: exceptionReason, userInfo: exceptionUserInfo)
+			exceptionHandler.handle(exception: exception)
+		}
+	}
+	
+	@available(iOS 13.0, *)
+	class DiffableDataSourceFunctionalTableDataImpl: FunctionalTableDataImpl {
+		var datasource: DiffableDataSource?
+		var isRendering: Bool = false
```
[the remark this answers is not in the document]
Comment (josephroquedev) [reply Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift:262]: Should we still make it true/false at the start/end of `renderAndDiff`, for posterity?

File: Sources/FunctionalTableData/CellConfigType.swift [hunk 6]
```diff
@@ -63,3 +63,72 @@ public extension CellConfigType {
...
+		get { return base.actions }
+		set { base.actions = newValue }
+	}
+	
+	var accessibility: Accessibility {
+		get { return base.accessibility }
+		set { base.accessibility = newValue }
+	}
+	
+	private var base: CellConfigType
```
[the remark this answers is not in the document]
Comment (jasonkemp9099) [reply Sources/FunctionalTableData/CellConfigType.swift:95]: Can't be `let` because CellConfigType is not constrained to objects and there are setters. Will make public.

File: Sources/FunctionalTableData/CellConfigType.swift [hunk 7]
```diff
@@ -63,3 +63,72 @@ public extension CellConfigType {
...
+	
+	func update(cell: UITableViewCell, in tableView: UITableView) {
+		base.update(cell: cell, in: tableView)
+	}
+	
+	func update(cell: UICollectionViewCell, in collectionView: UICollectionView) {
+		base.update(cell: cell, in: collectionView)
+	}
+	
+	func isEqual(_ other: CellConfigType) -> Bool {
```
[the remark this answers is not in the document]
Comment (jasonkemp9099) [reply Sources/FunctionalTableData/CellConfigType.swift:116]: Yeah, probably. Good call.

File: Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (jasonkemp9099) [reply Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift:262]: Fine. But I'm only doing it for approvals! :)

**Review by Campbell-Scott — APPROVED** [review]
Code & test updates looking good ✅ 
Manual testing also looking good ✅

**Review by josephroquedev — APPROVED** [review]

File: Sources/FunctionalTableData/CellConfigType.swift [hunk 8]
```diff
@@ -63,3 +63,80 @@ public extension CellConfigType {
...
 	}
 }
+
+public struct AnyCellConfigType: CellConfigType, Hashable {
+	public func dequeueCell(from tableView: UITableView, at indexPath: IndexPath) -> UITableViewCell {
+		return base.dequeueCell(from: tableView, at: indexPath)
+	}
+	
+	public static func ==(lhs: AnyCellConfigType, rhs: AnyCellConfigType) -> Bool {
+		return lhs.sectionKey == rhs.sectionKey && lhs.key == rhs.key && lhs.base.isSameKind(as: rhs.base) && lhs.base.isEqual(rhs.base)
```
Comment (Caiopia) [inline Sources/FunctionalTableData/CellConfigType.swift:73]: Any reason we can't we defer to the `isEqual` method here? I'd expect both `==` and `isEqual` methods to be equivalent.

```suggestion
		return lhs.isEqual(rhs)
```

Comment (jasonkemp9099) [reply Sources/FunctionalTableData/CellConfigType.swift:73]: In this case, no.

Equality and hash have to use the same fields, otherwise the diffable data source will sometimes fail to apply the snapshot (with an exception).

We include the sectionKey here to tie each cell with its section to account for a cell with the same key and state in multiple sections. If we don't, diffable data sources just uses the last one it saw with the same hash. I think it stores sections and rows together in the same hashtable or something.

`AnyCellConfigType` is solely around for being Hashable, to satisfy the compiler and the requirements of diffable data sources. It should probably be internal, but there's no harm in making it public. For its `CellConfigType` methods, it should defer to its `base` instance.

File: Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift [hunk 9]
```diff
@@ -57,4 +57,347 @@ extension FunctionalTableData {
 			return cellConfig.actions.hasEditActions || self.tableView(tableView, canMoveRowAt: indexPath) || cellConfig.style?.selected != nil
 		}
 	}
+	
+	class OldFunctionalTableDataImpl: FunctionalTableDataImpl {
```
Comment (Caiopia) [inline Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift:61]: Styling: Can we move some of these types to their own files to short this file up a bit?

File: Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift [hunk 10]
```diff
@@ -57,4 +57,347 @@ extension FunctionalTableData {
...
+				dataSource.defaultRowAnimation = .none
+				self.datasource = dataSource
+				
+			}
+		}
+		
+		init(name: String, cellStyler: CellStyler) {
+			self.name = name
+			self.cellStyler = cellStyler
+			self.applyQueue = DispatchQueue.global(qos: .background)
```
Comment (Caiopia) [inline Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift:321]: I imagine that snapshot applying is frequently tied to the user's interactions. Is there another queue with higher priority we could use? Maybe `.userInteractive`?

Comment (jasonkemp9099) [reply Sources/FunctionalTableData/TableView/FunctionalTableData+UITableViewDataSource.swift]: Deleted this. Wasn't using the queue because it broke tests and didn't notice improvements.

File: Tests/FunctionalTableDataTests/DiffableDataSourceFunctionalTableDataPerformanceTests.swift [hunk 11]
```diff
@@ -0,0 +1,79 @@
+//
+//  FunctionaltableDataPerformanceTests.swift
```
Comment (Caiopia) [inline Tests/FunctionalTableDataTests/DiffableDataSourceFunctionalTableDataPerformanceTests.swift:2]: Small detail:

```suggestion
//  DiffableDataSourceFunctionalTableDataPerformanceTests.swift
```

File: Sources/FunctionalTableData/TableView/FunctionalTableData+DiffableDataSource.swift [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+			let changeSet = TableSectionChangeSet(old: oldSections, new: localSections, visibleIndexPaths: indexPaths)
+			datasource.data.sections = localSections
+			
+			var snapshot = NSDiffableDataSourceSnapshot<DiffableTableSection, AnyCellConfigType>()
+			let diffableSections = localSections.map { DiffableTableSection($0) }
+			snapshot.appendSections(diffableSections)
+			for newSection in diffableSections {
+				snapshot.appendItems(newSection.anyRows, toSection: newSection)
+			}
+			var reallyAnimated: Bool = true
```
Comment (Caiopia) [inline Sources/FunctionalTableData/TableView/FunctionalTableData+DiffableDataSource.swift:90]: nit: Haha, these variable names are great, but maybe something like `isFirstRender` or `isFirstSnapshot` could be a bit more descriptive, followed by `let shouldAnimate = animated && !isFirstRender`?

**Review by Caiopia — APPROVED** [review]
LGTM - Thanks for your patience with this and awesome work! I'm excited for this.

For visibility - There is a known limitation right now with the new diffing strategy where it can't detect certain state changes that result in a cell reload. This is due to the hashable implementation being mostly key based. A current temporary workaround is to embed some state into the key so that diffable data source can detect a state change and reload. However, long-term, a `Hashable` requirement for view states should eventually be added so that changes can be detected regardless of key.
